### PR TITLE
Support serialization of geo shapes on documents

### LINKED ISSUE #3096: Geo-Shape - invalid serialization when used JsonNetSerializer.Default
**NEST/Elasticsearch.Net version**: 6.0.1

**Elasticsearch version**: 6.2.1

**Description of the problem including expected versus actual behavior**:

We are indexing the data containing geo-shape data.
When using ``` JsonNetSerializer``` these data are serialized incorrectly. 
(Default serializer works OK.)

```c#
var point = new PointGeoShape(new GeoCoordinate(lat, lon));
```
**With default serializer (expected):**
```json
{
    "geometry": {
        "type": "geometrycollection",
        "geometries": [
            {
                "coordinates": [
                    49.105995,
                    21.9926492
                ],
                "type": "point"
            },
            {
                "coordinates": [
                    49.1017987,
                    21.9909713
                ],
                "type": "point"
            }
        ]
    }
}
```

**With JsonNetSerializer (actual):**
```json
{
    "geometry": {
        "type": "geometrycollection",
        "geometries": [
            {
                "coordinates": {
                    "lat": 49.105995,
                    "lon": 21.9926492
                },
                "type": "point"
            },
            {
                "coordinates": {
                    "lat": 49.1017987,
                    "lon": 21.9909713
                },
                "type": "point"
            }
        ]
    }
}
```
which produces the following error:
```
json_parse_exception Reason: "Current token (FIELD_NAME) not numeric, can not use numeric value accessors
 at [Source: org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper@248dbc2c; line: 1, column: 462]
```

### LINKED ISSUE #3100: De-serialization of Geometry using NEST
**NEST/Elasticsearch.Net version**: 6.0.1

**Elasticsearch version**: 6.2.1

**Description of the problem including expected versus actual behavior**: I am trying to include GeometryCollection types within my POCO but de-serialization fails because it doesn't know how to de-serialize IGeoShape in the .Geometries enumerable property. I have tried with adn without the JsonNetSerializer to no avail.

**Steps to reproduce**:
 1. Create a POCO with a GeometryCollection property.
 2. Query an elastic search index for the result type.
 3. Note de-serialization error:

Exception thrown: 'Elasticsearch.Net.UnexpectedElasticsearchClientException' in Elasticsearch.Net.dll
An exception of type 'Elasticsearch.Net.UnexpectedElasticsearchClientException' occurred in Elasticsearch.Net.dll but was not handled in user code
Could not create an instance of type Nest.IGeoShape. Type is an interface or abstract class and cannot be instantiated. Path 'secondaryLocations.geometries[0].type', line 18, position 15.

**Provide `ConnectionSettings` (if relevant)**:

var node = new Uri("http://localhost:9200");

var pool = new SingleNodeConnectionPool(node);

var settings = new ConnectionSettings(pool)
                .DefaultMappingFor<Location>(m => m.IdProperty(doc => doc.Id)
                    .IndexName("locations")
                )
                .DisableDirectStreaming();

var client = new ElasticClient(settings);

**Provide `DebugInformation` (if relevant)**:

Exception thrown: 'Elasticsearch.Net.UnexpectedElasticsearchClientException' in Elasticsearch.Net.dll
An exception of type 'Elasticsearch.Net.UnexpectedElasticsearchClientException' occurred in Elasticsearch.Net.dll but was not handled in user code
Could not create an instance of type Nest.IGeoShape. Type is an interface or abstract class and cannot be instantiated. Path 'secondaryLocations.geometries[0].type', line 18, position 15.

{
    "category": {
        "id": "12345",
        "key": "distributer",
        "name": "Distributer"
    },
    "primaryLocation": {
        "type": "point",
        "coordinates" : [-77.03653,38.897676]
    },
    "secondaryLocations" : {
        "type": "geometrycollection",
        "geometries": [
            {
                "type": "point",
                "coordinates": [100.0, 0.0]
            },
            {
                "type": "linestring",
                "coordinates": [ [101.0, 0.0], [102.0, 1.0] ]
            }
        ]
    }
}

**Describe the feature**:

### PR DESCRIPTION
**Pull request by russcam**
Fixes #3100
Fixes #3096

### CODE REVIEW

**Comment by russcam** [comment]
I think the geo shape support should be revisited in 7.x. For example,

- `GeometryCollection` should implement `IGeoShape` as it is a valid input for a `geo_shape` field datatype.
- All geoshape queries could be condensed into one query

File: src/Serializers/Nest.JsonNetSerializer/Converters/HandleNestTypesOnSourceJsonConverter.cs [hunk 1]
```diff
@@ -40,14 +41,19 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
-		private static readonly Type[] NestTypesThatCanAppearInSource = {
+		private static readonly HashSet<Type> NestTypesThatCanAppearInSource = new HashSet<Type>
+		{
 			typeof(JoinField),
 			typeof(QueryContainer),
 			typeof(CompletionField),
 			typeof(Attachment),
-			typeof(ILazyDocument)
+			typeof(ILazyDocument),
+			typeof(GeoCoordinate)
```
Comment (codebrain) [inline src/Serializers/Nest.JsonNetSerializer/Converters/HandleNestTypesOnSourceJsonConverter.cs]: `IGeoShape` as well as `IGeometryCollection` ?

File: src/Tests/Framework/MockData/Shape.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			var points = new List<GeoCoordinate>();
+
+			for (var i = 0; i < count; i++)
+				points.Add(GenerateGeoCoordinate(p));
+
+			return points;
+		}
+
+		// adapted from https://gis.stackexchange.com/a/103465/30046
+		private static IEnumerable<GeoCoordinate> GeneratePolygonCoordinates(Faker p, GeoCoordinate centroid, double maxDistance = 0.0002)
```
Comment (codebrain) [inline src/Tests/Framework/MockData/Shape.cs]: :+1:

File: src/Serializers/Nest.JsonNetSerializer/Converters/HandleNestTypesOnSourceJsonConverter.cs (hunk 1 above)
Comment (russcam) [reply src/Serializers/Nest.JsonNetSerializer/Converters/HandleNestTypesOnSourceJsonConverter.cs]: Handled on lines 56 and 57 below. It's possible that a user could specify any type that implements `IGeoShape` e.g. `IPolygonGeoShape`, so json conversion should check if `IGeoShape` is assignable from the type.

**Pull request merged**